src/Reserve.sol

```diff
@@ -37,44 +122,146 @@ contract Reserve is IReserve, Ownable {
         _;
     }
 
-    function withdraw(
-        IERC20 token,
-        address to,
-        uint256 amt
-    ) public override onlyUser {
-        uint256 balance = withdrawable[token];
-        require(balance >= amt, "Reserve: withdrawal over balance");
-        withdrawable[token] = balance - amt;
-        require(token.transfer(to, amt), "Reserve: transfer failed");
-        emit Withdrawn(msg.sender, token, to, amt);
+    /// @notice Sets a plugin for a given token.
+    /// All future deposits and withdrawals of that token will be made using that plugin.
+    /// All currently deposited tokens of that type will be withdrawn from the plugin previously
+    /// set for that token and deposited into the new one.
+    /// If no plugin has been set, funds are deposited from the reserve itself.
+    /// If no plugin is being set, funds are deposited into the reserve itself.
+    /// Callable only by the current owner.
+    /// @param token The used token.
+    /// @param newPlugin The new plugin address. `NO_PLUGIN` if no plugin should be used.
+    function setPlugin(IERC20 token, IReservePlugin newPlugin) public onlyOwner {
+        IReservePlugin oldPlugin = plugins[token];
+        plugins[token] = newPlugin;
+        uint256 amt = deposited[token];
```

Comment:
I think in the case of a plugin migration we should not use `deposited` amount.

Instead, we should take all the balance (deposit + interest/earned amount). Otherwise the interest gets lost.

I think each plugin should have a common interface for it.
```solidity
plugin.balance() // or totalAvialable
```

Reply:
If we implement the `surplusHarvest` pattern. See other comment. We should call the `surplusHarvest` method before migrating to another reserve.